### PR TITLE
Fix copy as image for chinese

### PR DESCRIPTION
**Pull request by SlySven**
This is a collection of 5 commits that better handles the text that makes up each line of text in the `TBuffer::lineBuffer` specifically they iterate through the text on a grapheme by grapheme basis and considered the effective width of the glyphs from the passing of the **first** Unicode codepoint of the grapheme into the `getGraphemeWidth(ucs4)` method.  This means that non-BMP characters and combining diacriticals are now correctly accounted for - and stops the production of:
```
"TTextEdit::getGraphemeWidth(...) WARN - trying to get width of a Unicode character which is unprintable, codepoint number: U+xxxx"
```
error messages when the individual surrogates were previously pushed into that method and:
```
"TTextEdit::getGraphemeWidth(...) WARN - trying to get width of a Unicode character which is a zero width combiner, codepoint number: U+xxxx"
```
when combining diacriticals were so pushed...!

It also switches to using `QFontMetrics::averageCharWidth()` instead of trying to measure the width of semi-randomly chosen `QChars` i.e. space and 'W' with `QFontMetrics::width(QChar)` which has been declared ***obsolete*** and returns a potentially seg. faulting `0` in some situations that we *have* been encountering - see #2391 .

### CODE REVIEW

File: src/TTextEdit.cpp [hunk 1]
```diff
@@ -445,20 +445,26 @@ inline uint TTextEdit::getGraphemeBaseCharacter(const QString& str) const
...
+
     QChar first = str.at(0);
     if (first.isSurrogate() && str.size() >= 2) {
         QChar second = str.at(1);
         if (first.isHighSurrogate() && second.isLowSurrogate()) {
             return QChar::surrogateToUcs4(first, second);
-        } else if (first.isLowSurrogate() && second.isHighSurrogate()) {
+        }
+
+        // Although this has been coded for it is not what I expect to see - SlySven:
```
Comment (vadi2) [inline src/TTextEdit.cpp]: The `Q_UNLIKELY` says enough that this is not to be expected, so the comment doesn't add value - let's have less text the programmer needs to see. We've got thousands of lines of code already!

File: src/TConsole.cpp [hunk 2]
```diff
@@ -1120,7 +1120,7 @@ void TConsole::changeColors()
         const QString t = "123";
         p.drawText(r, 1, t, &r2);
         // N/U:        int mFontHeight = QFontMetrics( mDisplayFont ).height();
-        int mFontWidth = QFontMetrics(mDisplayFont).width(QChar('W'));
+        int mFontWidth = QFontMetrics(mDisplayFont).averageCharWidth();
```
Comment (vadi2) [inline src/TConsole.cpp:1123]: Unrelated change... should be a separate PR. Scope creep is not good!

File: src/TTextEdit.cpp [hunk 3]
```diff
@@ -983,37 +989,77 @@ void TTextEdit::mouseMoveEvent(QMouseEvent* event)
     highlight();
 }
 
+// Returns the index into the relevant TBuffer::lineBuffer of the FIRST QChar
+// of the grapheme under the mouse - it ALSO returns zero (which will probably
+// NOT be a valid index) if there is no valid index to return; it might be
```
Comment (vadi2) [inline src/TTextEdit.cpp:993]: Zero is the first character in the line, that is a valid index.

File: src/TTextEdit.cpp [hunk 4]
```diff
@@ -983,37 +989,77 @@ void TTextEdit::mouseMoveEvent(QMouseEvent* event)
...
-        int characterWidth = 1;
+#if defined (QT_DEBUG)
+    // Enable this from debugger to show output
+    static bool debug_mousePositionInLine = false;
+#endif
+
+    if (lineNumber >= 0 && lineNumber < mpBuffer->lineBuffer.size()) {
+        // Line number is (should be) within range of lines in the
+        // TBuffer::lineBuffer - might need to check that this still works after
+        // that buffer has reached the limit when it starts to have the
```
Comment (vadi2) [inline src/TTextEdit.cpp:1001]: Did you check this?

File: src/TTextEdit.cpp [hunk 5]
```diff
@@ -1397,6 +1443,9 @@ void TTextEdit::slot_copySelectionToClipboardHTML()
     forceUpdate();
 }
 
+// Technically this copies whole lines into the image even if the selection does
+// not start at the beginning of the first line or end at the last grapheme on
+// the last line.
```
Comment (vadi2) [inline src/TTextEdit.cpp:1425]: Not a useful code comment (good piece of trivia, though :wink:)

File: src/TTextEdit.cpp [hunk 6]
```diff
@@ -1431,27 +1480,43 @@ void TTextEdit::slot_copySelectionToClipboardImage()
     }
 
     // Qt says: "Maximum supported image dimension is 65500 pixels" in stdout
-    auto heightpx = qMin(65500, (mPB.y() - mPA.y() + 1) * mFontHeight);
+    auto heightpx = std::min(65500, (mPB.y() - mPA.y() + 1) * mFontHeight);
```
Comment (vadi2) [inline src/TTextEdit.cpp:1460]: :+1:

File: src/TTextEdit.cpp [hunk 7]
```diff
@@ -983,37 +989,77 @@ void TTextEdit::mouseMoveEvent(QMouseEvent* event)
...
+        int column = 0;
+        // This is the calculated (estimated) cumulative width of all of the
+        // graphemes considered in the line so far:
         int currentX = 0;
-        for (const auto& character : mpBuffer->lineBuffer.at(lineNumber)) {
-            const uint unicode = getGraphemeBaseCharacter(character);
+        QString lineText = mpBuffer->lineBuffer.at(lineNumber);
+        QTextBoundaryFinder boundaryFinder(QTextBoundaryFinder::Grapheme, lineText);
+#if defined (QT_DEBUG)
+        if (debug_mousePositionInLine) {
```
Comment (vadi2) [inline src/TTextEdit.cpp]: These sprinkled debug statements make the code hard to read. If the algorithm is finished, lets clean up and remove the debugging scaffolding that was useful during development. I don't foresee us having to debug this often because once we get our code right, we don't need to do it again. This is unlike telnet negotiations and etc where the other party can mess things up.

Keeping this debug output around just makes it harder to understand the debug algorithm while it is trivial - 2mins of work - to re-add it when necessary.

File: src/TTextEdit.cpp [hunk 8]
```diff
@@ -983,37 +989,77 @@ void TTextEdit::mouseMoveEvent(QMouseEvent* event)
...
+            indexOfChar = nextBoundary;
         }
     }
 
-    characterIndex = mouseX / mFontWidth;
-    if (mShowTimeStamps) {
-        characterIndex -= 13;
-    }
-    characterIndex = std::max(characterIndex, 0);
-    return characterIndex;
```
Comment (vadi2) [inline src/TTextEdit.cpp:1016]: Why was all of this deleted? This keeps the old behaviour of still returning a theoretical X position for when you're out of the buffer range. I don't know how the code will behave when it's suddenly getting `0` now, have you tested and verified all of this?

File: src/TTextEdit.cpp [hunk 9]
```diff
@@ -1431,27 +1480,43 @@ void TTextEdit::slot_copySelectionToClipboardImage()
     }
 
     // Qt says: "Maximum supported image dimension is 65500 pixels" in stdout
-    auto heightpx = qMin(65500, (mPB.y() - mPA.y() + 1) * mFontHeight);
+    auto heightpx = std::min(65500, (mPB.y() - mPA.y() + 1) * mFontHeight);
     auto lineOffset = mPA.y();
 
     // find the biggest width of text we need to work with
     int largestLine{};
+    qDebug().noquote().nospace() << "TTextEdit::slot_copySelectionToClipboardImage() INFO - line widths in \"normal\" character widths (and pixels):";
```
Comment (vadi2) [inline src/TTextEdit.cpp]: This debug information is not necessary, I guess this is a testing left over?

File: src/TTextEdit.cpp [hunk 10]
```diff
@@ -1431,27 +1480,43 @@ void TTextEdit::slot_copySelectionToClipboardImage()
...
-            lineWidth += characterWidth;
+            column +=charWidth;
+            // The timestamp is (currently) 13 "normal width" characters
+            // but that might not always be the case in some future I18n
+            // situations:
+            lineWidth = (mShowTimeStamps ? 13 + column : column) * mFontWidth;
+            // Increment for loop:
+            indexOfChar = nextBoundary;
         }
+        qDebug().noquote().nospace() << "\"" << lineText << "\" " << column << " (" << lineWidth << ")";
```
Comment (vadi2) [inline src/TTextEdit.cpp]: Same, not necessary debug information.

File: src/TTextEdit.cpp [hunk 11]
```diff
@@ -983,37 +989,77 @@ void TTextEdit::mouseMoveEvent(QMouseEvent* event)
...
+                    // The timestamp is (currently) 13 "normal width" characters
+                    // but that might not always be the case in some future I18n
+                    // situations:
+                    return std::max(0, indexOfChar - 13);
                 }
-                characterIndex = std::max(characterIndex, 0);
-                return characterIndex;
+                return std::max(0, indexOfChar);
             }
+            // Increment for loop:
```
Comment (vadi2) [inline src/TTextEdit.cpp]: This is not confusing nor difficult to understand - please, please let's not have such trivial comments such as "increment for loop".

**Review by vadi2 — CHANGES_REQUESTED** [review]
after

This PR breaks text selection, has scope creep, adds debug statements that make the actual code harder to read, as well as a unnecessary comments that _also_ add to the overhead of working with code. I'm not really happy about this, sorry...

File: src/TTextEdit.cpp (hunk 1 above)
Comment (SlySven) [reply src/TTextEdit.cpp]: Someone else had put those lines in, I don't *think* they are necessary, but I am not certain enough that I want to remove them...

File: src/TTextEdit.cpp (hunk 3 above)
Comment (SlySven) [reply src/TTextEdit.cpp:993]: Not if the line is an empty one! We do not include the EOL character and we do get lines with NO content - such that using `TBuffer::lineBuffer.at(row).at(column)` on them is NOT valid and will seg. fault if tried... :cry:

File: src/TTextEdit.cpp (hunk 4 above)
Comment (SlySven) [reply src/TTextEdit.cpp:1001]: No - though I am *aware* that we do have an open issue about things breaking when the buffer has reached the set limit and has started to truncate itself (I haven't been able to find it though which is a pain).

I am not certain at this time how that process proceeds and with so many balls in the air right now I haven't taken the time to try and get my head around it - I'm just hoping it isn't going to surprise me/us... :pray:

File: src/TTextEdit.cpp (hunk 5 above)
Comment (SlySven) [reply src/TTextEdit.cpp:1425]: It is useful - and it points out a disparity in the behaviour compared to what might be anticipated from assigning values to `mPA::x` and `mPB::x` - as I said technically the disparity makes sense - as forming a rectangular picture *but blanking off the parts of the first and last line outside of the selection* would be even harder to do right for the image form than it was for the text/HTML case - and remember how mixed the feelings about *that* were!

File: src/TTextEdit.cpp (hunk 7 above)
Comment (SlySven) [reply src/TTextEdit.cpp]: Okay - will nuke it - it was very useful to determine the grapheme that was believed to be the one under the cursor but it was *exceptionally* spammy (as it runs for ANY mouse movement over the `TConsole`)...

File: src/TConsole.cpp (hunk 2 above)
Comment (SlySven) [reply src/TConsole.cpp:1123]: For things to work as expected it is important that *all* measurements of text widths use the same basis - some using spaces and some using `W` is **not** consistent - so I do not believe this is Scope creep...

File: src/TTextEdit.cpp (hunk 1 above)
Comment (SlySven) [reply src/TTextEdit.cpp]: ... and putting my name there indicates who should be contacted if it ever ***is*** found.

File: src/TTextEdit.cpp (hunk 11 above)
Comment (SlySven) [reply src/TTextEdit.cpp]: It was also a reminder for me to **ensure** that I put it in after noting that I was NOT doing it in the (missing third part of the) `for (..,; ...; ...)` ! AFAICR It needs to be done in that way in order to sequences the updating of the start and end `QChar` indexes so that they correctly select the right ones to form a *single* grapheme.

File: src/TConsole.cpp (hunk 2 above)
Comment (vadi2) [reply src/TConsole.cpp:1123]: In a monospace font, it is 😅 but don't take it out now.

File: src/TTextEdit.cpp (hunk 1 above)
Comment (vadi2) [reply src/TTextEdit.cpp]: That's not how people will see it and I don't think it adds value. Let's keep code minimal with only the necessary explanations.

File: src/TTextEdit.cpp (hunk 11 above)
Comment (vadi2) [reply src/TTextEdit.cpp]: Ok, but it's in now, we don't need it anymore.

File: src/TTextEdit.cpp (hunk 8 above)
Comment (SlySven) [reply src/TTextEdit.cpp:1016]: If the `lineNumber` argument is out of range then there is **NO** useful value to return! The "BufferX" in the method name is not a spatial characteristic but an array index - note that the value I return is based on `indexOfChar` which is the first `QChar` in the particular grapheme under the mouse cursor on the relevant line {perhaps that should have been `indexOfQChar`} and it would not be possible to derive a meaningful value in the absence of any graphemes to draw and estimate their width from.

*Perhaps* that is not entirely fair: it *might* be possible to estimate the number of *average* width characters - given that that is how much space we now assign for a single, "normal" width grapheme - but to than take that number and to use it to index into a random line of text in the `TBuffer::lineBuffer` would be *nonsensical* even if it did not cause a seg. fault by trying to access beyond the end of that random line.

Thus in the situation when it is not possible to access the buffer either we could return `-1` as a sentinel value as I hypothesize or **we return `0` (as we do currently) and rely on the caller to check that `lineNumber` of the buffer will support the index value returned**...

File: src/TTextEdit.cpp (hunk 9 above)
Comment (SlySven) [reply src/TTextEdit.cpp]: Yeah - and it did lead me to a bug I had made - it can probably be gotten rid of along with the other things...

File: src/TTextEdit.cpp (hunk 8 above)
Comment (vadi2) [reply src/TTextEdit.cpp:1016]: Ok but that's how the rest of the code works right now and I don't want to refactor it entirely and or introduce a behaviour change. It's one day until feature freeze, now is not the time

File: src/TTextEdit.cpp (hunk 10 above)
Comment (SlySven) [reply src/TTextEdit.cpp]: Yeah.

File: src/TTextEdit.cpp (hunk 8 above)
Comment (SlySven) [reply src/TTextEdit.cpp:1016]: I've taken a closer look at b3f37615ed41306393da55d1d1df7e061762f2da and what you are saying about "*that is how the rest of the code works right now*" seems to confirm to me that that is why the existing code *doesn't* work for non-BMP and combining diacriticals precisely because there is *not* a 1:1 relationship between how far across from the left margin the mouse X position is and the index into the `QChar`s that form the `QString` that is the textual data for each line of Mud text.  By 1:1 I mean that for a certain number of pixels from the left side the index of the `QChar` that is at that number of pixels is a fixed and invariant value. That is what the previous code assumed and it is only true for BMP characters without any diacriticals (and without any non-printing Unicode codepoints such as Variation Selectors, Joiners, etc.)

The code I am proposing is good enough provided that we assume that the text we present (barring tabs) can be done in a duo-spaced (each glyph takes up either one or two "spaces") manner. {As it happens I have prototyped code that does things in a slower and more involved manner that actually recorded the width each grapheme took when painted on screen and looked that up when it need to work out where the mouse movement put the cursor - which meant it could handle Qt using glyphs from different fonts as needed to display the text - which this code will not cope with too well if the individual characters deviate *too* much from the "average" width.}

Saying "it's is one day until feature freeze" is not helpful - either we want code that works for a non-English user base or we put forward a product that has known miss-features.

I am sorry if I did not point this out more strongly when the 4.0.0. release was recently discussed but as I said at the time I did not feel we had everything sorted yet and I pointed out the limits on my current attention to the project and the reason why. It was only when I started looking at your PR - that this one is hacking away at - that I realised that I hadn't sorted out the selection and highlighting code as completely as I had intended. OTOH This will, I believe, be good enough for Western and Chinese languages cases.

File: src/TTextEdit.cpp (hunk 1 above)
Comment (SlySven) [reply src/TTextEdit.cpp]: Okay, I will remove comment (and revise the debug text).

File: src/TTextEdit.cpp (hunk 11 above)
Comment (SlySven) [reply src/TTextEdit.cpp]: Okay, will be gone in next commit.

File: src/TTextEdit.cpp (hunk 6 above)
Comment (SlySven) [reply src/TTextEdit.cpp:1460]: I spotted another one that I had not changed *within the range of my modifications* - will change that in next commit as well. :blush:

File: src/TTextEdit.cpp (hunk 9 above)
Comment (SlySven) [reply src/TTextEdit.cpp]: ... will be gone in next commit.

File: src/TTextEdit.cpp (hunk 10 above)
Comment (SlySven) [reply src/TTextEdit.cpp]: It is banished in next commit.

File: src/TTextEdit.cpp (hunk 4 above)
Comment (SlySven) [reply src/TTextEdit.cpp:1001]: > ... though I am aware that we do have an open issue about things breaking when the buffer has reached the set limit and has started to truncate itself ...

There **is** something screwy going on in that situation - if the display is NOT split the on-screen display seems to update a few lines at the bottom but then redraws the same lines and you do not see the new content that - if the `TConsole` *is* split does scroll up in the lower pane. The bottom line is that this new code does not *seem* to be any *more* broken by this than the previous code - OTOH there is definitely something buggy there but what on earth it is is not clear to me at the present... for reference I used a call of `setConsoleBufferSize(200,10)` to provoke that bug in the main console.

**Comment by SlySven** [comment]
The macOs Travis CI builds errored & failed (one of each) but the Linux builds all succeeded. For some reason though - I do not seem to have the permissions to re-request the unsuccessful CI jobs to be restarted. :confused:

File: src/TTextEdit.cpp (hunk 8 above)
Comment (vadi2) [reply src/TTextEdit.cpp:1016]: > That is what the previous code assumed and it is only true for BMP characters without any diacriticals (and without any non-printing Unicode codepoints such as Variation Selectors, Joiners, etc.)

I'm sorry but I don't follow what are you talking about. We've ruled out emojis as not a 4.0 requirement, so what usecases are you looking to fix here?

My original PR fixes it for Chinese, all the other languages we know of work fine. As far as I know there are no Vietnamese MUDs so that is not a concern. Is the concern you're having practical or theoretical?

**Review by vadi2 — APPROVED** [review]
